Implemention of new Poweradmin-Installer

SebTM opened the pull request:
1. Using Doctrine DBAL
2. PHP-Templates
3. Twitter-Bootstrap (currently from CDN)

fixes #230 - The Installer mentions that PowerDNS + default SQL-Schema needs to be installed. It also checks if tables from PowerDNS-Schema exists.

fixes #163 - Step 2

fixes #86 - Implemented third DNS-Server. We need to check where the variables gets used and add the new one.

fixes #77 - For me there is no need to provide sql-dumps for every database-schema. We have more work to maintain - this time should be used for writing / maintaining update/installer-script

SebTM commented:
**ToDo**

1. Implement a Error-/Exception-Handler class to improve the way of throwing and handling errors. For me the current way is awful so I skipped implementing it and do it directly right!
2. Need someone who reviews my English phrases and correct them before generating translations-template
3. Need testers - especially for PostgreSQL, if someone have time and could intensive test SQLite it would be great too. MySQL/MariaDb is fully working as it is my primary platform.

@edmondas Here you can review my changes :) Can you help me resolving the merge-issues? I tried to rebase on current master but it seems to not solve it.

edmondas commented on .gitignore:
```diff
@@ -1,10 +1,3 @@
+/.idea/
 inc/config.inc.php
-.DS_Store
```
please do not remove old git ignore entries

SebTM replied:
There were so many entries for me without function. Would be good to comment them with their sense. Do we only need the .DS_Store back or all?

edmondas replied:
I can comment all of them:

-/nbproject - ignore NetBeans stuff, let's keep it for user that will work with this IDE
 -/scripts - ignore scripts from https://github.com/poweradmin/poweradmin-scripts that I use from time to time to update TLDs or create release
 -/tests - tests that I used to test current stability of code (https://github.com/poweradmin/poweradmin-tests), leave it
 -/docs - this can be removed
 -/.settings/ - Eclipse? I used it once, but let's leave it
 -/.buildpath - not sure what it is, remove.
 -/.project - Eclipse? leave it.

SebTM replied:
Fixed

edmondas commented on inc/classes/Db.php:
```diff
@@ -0,0 +1,117 @@
+    public static function createConnection($parameters, $name = 'default')
+    {
+        // Check for driver in parameters
+        if (!is_array($parameters) || !array_key_exists('driver', $parameters)) {
```
I think connection parameter checking can be extracted to separate method and it can be private method which is called before making connection or it would be better to have it public and use it before calling createConnection() in our code, that way we could handle and respond to exception in other parts of code (for example controllers).

So it will look smth like this:
```
// UI controller
if (!Db::validateParameters($parameters) ) {
 // handle parameters issue, display error, stop execution here
}

Db::createConnection($parameters)
// handle connection availability here
```

SebTM replied:
Implemented method, in this case it will not replace form validation but its another security-step.

edmondas commented on inc/database.inc.php:
```diff
@@ -67,7 +67,6 @@ function dbConnect() {
     global $db_charset;
     global $db_file;
     global $db_debug;
-    global $db_ssl_ca;
```
It is strange that it's not used anymore, probably someone removed it previously, doh...

SebTM replied:
It was supported here but only for mysqli: https://github.com/poweradmin/poweradmin/blob/9731ac8dfe42f67d011058f486f7fb3a9b01d67f/inc/database.inc.php

The possibility also exists with Doctrine DBAL but I don't see a need to implement it as it's very specific. (http://stackoverflow.com/questions/19039947/symfony2-doctrine-connect-to-database-via-ssl)

edmondas commented on inc/plugins/auth_local/auth_local.plugin.php:
```diff
@@ -29,7 +29,6 @@
  * @copyright   2010-2014 Poweradmin Development Team
  * @license     http://opensource.org/licenses/GPL-3.0 GPL
  */
-require_once dirname(dirname(dirname(__DIR__))) . '/vendor/poweradmin/Password.php';
```
Have you checked if it really works after removing it?

SebTM replied:
I've checked that it's working. I included the autoloader provided by composer inside the toolkit.inc.php

edmondas commented on index.php:
```diff
@@ -21,6 +21,9 @@
  */
 
+error_reporting(E_ALL);
+ini_set('display_errors', 1);
```
this is good for development but should be removed for release

SebTM replied:
Oh, I missed this! Removed in last commit.

edmondas commented on install/htaccess.dist:
```diff
@@ -1,5 +0,0 @@
-<IfModule mod_rewrite.c>
```
maybe we can move this to the top directory but not remove?

SebTM replied:
Shouldn't we check first, that it still follows the requirements for DynDNS? Or are there no standards?

edmondas commented on install/includes/database_structure.php:
```diff
@@ -0,0 +1,94 @@
+    'perm_items' => [
+        'columns' => [
+            'id' => ['type' => 'integer', 'options' => ['autoincrement' => true ]],
+            'name' => ['type' => 'string'],
```
what will be the length of this field? Same goes to other fields which type is string.

SebTM replied:
string goes varchar with 255

edmondas commented on install/includes/functions.php:
```diff
@@ -0,0 +1,114 @@
+            $userData['username'], $userData['password'], $userData['fullname'], $userData['email'],
+            'Administrator with full rights.', 1, false, true
+        ]
+    );
```
Don't you forgot to insert admin permission for this user? Previously it was:
```
INSERT INTO perm_templ_items (templ_id, perm_id) VALUES (1,53)
```

edmondas replied:
Ok, found this in another place.

SebTM replied:
Yeah, it gets created more up after importing permissions.

edmondas commented on install/includes/validation_functions.php:
```diff
@@ -0,0 +1,13 @@
+        return true;
+    }
+
+    return false;
```
why not simply return is_file($value) ?

SebTM replied:
You are right, fixed!

edmondas commented on install/index.php:
```diff
@@ -1,276 +1,419 @@
+session_start();
+
+// Set directory-constants
+define('INSTALLER_DIRECTORY', __DIR__ . '/');
```
add define('TOP_DIRECTORY', dirname(__DIR__);

SebTM replied:
fixed!

edmondas commented on install/index.php:
```diff
@@ -1,276 +1,419 @@
+define('INSTALLER_DIRECTORY', __DIR__ . '/');
+
+// Load autoloader & functions-file
+require_once INSTALLER_DIRECTORY . '../vendor/autoload.php';
```
require_once TOP_DIRECTORY . 'vendor/autoload.php';

SebTM replied:
fixed!

edmondas commented on install/index.php:
```diff
@@ -1,276 +1,419 @@
+    'supportedDatabaseDrivers' => ['pdo_mysql' => 'MySQL', 'pdo_pgsql' => 'PostgreSQL', 'pdo_sqlite' => 'SQLite'],
+
+    'defaultLocale' => 'en_US',
+    'locales' => ['de_DE' => _('German'), 'en_US' => _('English')],
```
Can we reuse some translations for other languages that we had previously?

SebTM replied:
Hmm, I think only partial the words. The phrases are completely rewritten. As long-term goal we should think about changing the translation-management to store the available locales in databases so the user can install the one he needs.

I've created a organization and project on Transifex. (https://www.transifex.com/poweradmin/)
If you tell me your Username I will add you as Administrator. I think we should wait until someone offers to maintain a language and then add it on Transifex.

edmondas commented on install/index.php:
```diff
@@ -1,276 +1,419 @@
+        $generalData = $_SESSION['generalData'];
+
+        // Replace placeholder
+        $configFileContent = str_replace([
```
maybe some parts of code can be extracted to includes/functions.php ? For example building of configuration file or creating db tables, it would be good to have less code in installer steps

SebTM replied:
I would leave it as it is as we will change the whole database-handling to ORM in the next versions. Also configuration-files (format and handling) should be changed as example to YAML with Symfony-Component to read/write and manage files.

edmondas commented on install/templates/step1.php:
```diff
@@ -0,0 +1,35 @@
+<?php include __DIR__ . '/header.php'; ?>
+
+    <h2><?= _('Welcome to the installation of Poweradmin!') ?></h2>
```
I'm just wondering if short php tags are enabled by default in various Linux distributions?

SebTM replied:
http://php.net/manual/en/language.basic-syntax.phptags.php
Since PHP 5.4 this tag is always available regardless of the short_open_tag ini setting.

edmondas commented:
Generally I like these changes, maybe cleanup a bit and extract some big parts of code to methods/functions, to simplify reading and understanding of code.

edmondas commented on install/includes/functions.php:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    // Check for error -> locale and even default locale not exist
+    if (!$result) {
```
I would suggest you to check also for default locale, because I don't have en_US.UTF-8 on my pc :)
if (!$result && $locale !== $defaultLocale) {

SebTM replied:
The default locale automatically gets checked. First we set the user locale - if failed - we check if default locale is not different to user locale and set it. Afterwards $result from user local - if executed default locale gets checked.

edmondas replied:
It simply fails with message: Error: The locale "en_US" does not exist!
And this will happen to other user as well. Another issue that I found with this English locale is:
new installer creates config with en_US, but there's en_EN directory in locale directory, also
selection of default English language fails en_US != en_EN.

edmondas replied:
```
$result = setlocale(LC_ALL, $locale);
```
returns false for me.
```
$ locale -a | grep en_US
en_US.utf8
```
So that means that default locale must be with utf suffix.
'defaultLocale' => 'en_US.utf8'

Could it be better not to stop user at first step, I would suggest to not use setlocale if default language 'english' is used.

edmondas commented on install/templates/step2.php:
```diff
@@ -0,0 +1,70 @@
+    <a href="<?= $_SERVER['PHP_SELF']; ?>?step=3<?=(defined('DEBUG') ? '&amp;debug=true' : '') ?>" class="btn btn-primary pull-right"><?= _('Continue'); ?></a>
+<?php elseif (!$createConfigFile || !$configFileIsWritable): ?>
+    <div
+        class="alert alert-danger"><?= sprintf(_('Could not create Config-File "%s" or it is not writable!'), basename($config['config_file'])); ?></div>
```
I think it should use $config['configFile'] not $config['config_file']

edmondas replied:
Also from this message it's not clear where exactly installer tries to create this file (maybe it should have inc/ prefix)?

SebTM replied:
You are right! I renamed the key to have a similar code-style inside the installer-files. I changed it to display the full path to config-file for now.

edmondas commented:
Btw Doctrine brings ~1000 files. Do we really need such big library? :)
I'm still considering Laravel for our next version, it too many things that we need:
mvc, templating and smaller orm that Doctrine.

edmondas commented:
Installer doesn't work properly with PostgreSQL:

_[Thu Mar 17 08:47:52.495307 2016] [:error] [pid 7552] [client 127.0.0.1:31958] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for type boolean: ""' in /home/edmondas/OpenSource/poweradmin/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91\nStack trace:\n#0 /home/edmondas/OpenSource/poweradmin/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php(91): PDOStatement->execute(Array)\n#1 /home/edmondas/OpenSource/poweradmin/vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(830): Doctrine\\DBAL\\Driver\\PDOStatement->execute(Array)\n#2 /home/edmondas/OpenSource/poweradmin/install/index.php(368): Doctrine\\DBAL\\Connection->executeQuery(Object(Doctrine\\DBAL\\Query\\QueryBuilder), Array)\n#3 {main}\n\nNext exception 'Doctrine\\DBAL\\Driver\\PDOException' with message 'SQLSTATE[22P02]: Invalid text representation: 7 ERROR:  invalid input syntax for type boolean: ""' in /home/edmondas/OpenSource/poweradmin/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:93\nStack trace:\n#0 /home/edmondas/OpenSou in /home/edmondas/OpenSource/poweradmin/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractPostgreSQLDriver.php on line 91_

edmondas commented:
New installer also do not work with SQLite:
```
Could not connect to database-server with the given informations!
```
Also I'm not sure why do you need to provide credentials for it :)
Another thing is that if there is error with DB settings then I need to select database type again, because it resets to MySQL.

SebTM commented:
I was choosing Doctrine as I like it and used it before using Symfony 2. If there is a better Solution e.g. with Laravel we can have a try. Is there the possibility to install Laravel DBAL-System and later ORM without the whole framework?

I will have a look at PostgreSQL the next days - didn't have time until now.

I'm searching for help regarding the problem with SQLite as there is a unsupported type error by doctrine using a db-file with default PoweDNS-Schema. (http://stackoverflow.com/questions/36073343/doctrine-dbal-type-boolean-for-sqlite/36073492#36073492)

SebTM commented:
As currently there were no real helpful post on Stackoverflow concerning the problem with doctrine I opened a ticket in issue-tracker of doctrine on Github. (https://github.com/doctrine/dbal/issues/2338)

SebTM commented:
Fixed Problem with translation-system (needed ".UTF-8"), added doctrine type-mapping so it's working with SQLite now and should be working with PostgreSQL. Could you check @edmondas ?

SebTM commented:
Updated Installed see 8f47b97 for details. Tested with all database-drivers. @edmondas Can you review please?

SebTM commented:
@JonasGroeger From my personal view I would also like to not add them to the repository. The issue at the creation time was that there is no build-workflow (as it was not needed) so we would need to introduce this to archive this.

SebTM commented:
@edmondas I've seen the installer has changed since the PR a lot and that there is recent action on the PR, can I help you with something on this?

edmondas commented:
@SebTM You made some significant changes, and I am integrating them bit by bit. This PR will be closed once the relevant parts have been added to the codebase.
